ext/crypto/00_crypto.js

```diff
@@ -955,9 +1065,147 @@
 
               return key;
             }
+            case "jwk": {
+              // 2. https://w3c.github.io/webcrypto/#rsassa-pkcs1-operations
+              let hash;
+
+              const jwk = keyData;
+              // 2.
+              if (
+                ArrayPrototypeFind(
+                  keyUsages,
+                  (u) =>
+                    !ArrayPrototypeIncludes([jwk.d ? "sign" : "verify"], u),
+                ) !== undefined
+              ) {
+                throw new DOMException("Invalid key usages", "SyntaxError");
+              }
+              // 3.
+              if (jwk.kty !== "RSA") {
+                throw new DOMException(
+                  "`kty` member of JsonWebKey must be `RSA`",
+                  "DataError",
+                );
+              }
+
+              // 4.
+              if (keyUsages.length > 0 && jwk.use && jwk.use !== "sig") {
+                throw new DOMException(
+                  "`use` member of JsonWebKey must be `sig`",
+                  "DataError",
+                );
+              }
+
+              // 5.
+              // Section 4.3 of RFC7517
+              if (jwk.key_ops) {
+                if (
+                  ArrayPrototypeFind(
+                    jwk.key_ops,
+                    (u) => !ArrayPrototypeIncludes(recognisedUsages, u),
+                  ) !== undefined
+                ) {
+                  throw new DOMException(
+                    "`key_ops` member of JsonWebKey is invalid",
+                    "DataError",
+                  );
+                }
+
+                if (
+                  !ArrayPrototypeEvery(
+                    jwk.key_ops,
+                    (u) => ArrayPrototypeIncludes(keyUsages, u),
+                  )
+                ) {
+                  throw new DOMException(
+                    "`key_ops` member of JsonWebKey is invalid",
+                    "DataError",
+                  );
+                }
+              }
+
+              // 6.
+              if (jwk.ext === false && extractable == true) {
+                throw new DOMException(
+                  "`ext` member of JsonWebKey is invalid",
+                  "DataError",
+                );
+              }
+
+              // 7.
+              if (jwk.alg) {
+                switch (jwk.alg) {
+                  case "RS1": {
+                    hash = "SHA-1";
+                    break;
+                  }
+                  case "RS256": {
+                    hash = "SHA-256";
+                    break;
+                  }
+                  case "RS384": {
+                    hash = "SHA-384";
+                    break;
+                  }
+                  case "RS512": {
+                    hash = "SHA-512";
+                    break;
+                  }
+                  default:
+                    throw new DOMException(
+                      "`alg` member of JsonWebKey invalid",
```

Comment:
```suggestion
                      "Hash algorithm not supported",
```

This can be extended by other specifications.

Reply:
ok, expanded to all instances, PS/RS/RSA-OAEP